livekit/src/room/participant/local_participant.rs

```diff
@@ -490,4 +604,196 @@ impl LocalParticipant {
     pub fn kind(&self) -> ParticipantKind {
         self.inner.info.read().kind
     }
+
+    pub async fn perform_rpc(
+        &self,
+        destination_identity: String,
+        method: String,
+        payload: String,
+        response_timeout_ms: Option<u32>,
+    ) -> Result<String, RpcError> {
+        let response_timeout = Duration::from_millis(response_timeout_ms.unwrap_or(10000) as u64);
+        let max_round_trip_latency = Duration::from_millis(2000);
+
+        if payload.len() > MAX_PAYLOAD_BYTES {
+            return Err(RpcError::built_in(RpcErrorCode::RequestPayloadTooLarge, None));
+        }
+
+        let id = Uuid::new_v4().to_string();
+        let (ack_tx, ack_rx) = oneshot::channel();
+        let (response_tx, response_rx) = oneshot::channel();
+
+        match self
+            .publish_rpc_request(
+                destination_identity.clone(),
+                id.clone(),
+                method.clone(),
+                payload.clone(),
+                (response_timeout - max_round_trip_latency).as_millis() as u32,
+                1,
+            )
+            .await
+        {
+            Ok(_) => {
+                self.local.pending_acks.lock().insert(id.clone(), ack_tx);
+                self.local.pending_responses.lock().insert(id.clone(), response_tx);
+            }
+            Err(e) => {
+                log::error!("Failed to publish RPC request: {}", e);
+                return Err(RpcError::built_in(RpcErrorCode::SendFailed, Some(e.to_string())));
+            }
+        }
+
+        // Wait for the ack first
+        let ack_result = tokio::time::timeout(max_round_trip_latency, ack_rx).await;
+
+        match ack_result {
+            Ok(_) => {
+                // Ack received, now wait for the response
+                match tokio::time::timeout(response_timeout, response_rx).await {
+                    Ok(response) => match response {
+                        Ok(result) => match result {
+                            Ok(payload) => {
+                                if payload.len() > MAX_PAYLOAD_BYTES {
+                                    Err(RpcError::built_in(
+                                        RpcErrorCode::ResponsePayloadTooLarge,
+                                        None,
+                                    ))
+                                } else {
+                                    Ok(payload)
+                                }
+                            }
+                            Err(e) => Err(e),
+                        },
+                        Err(_) => {
+                            Err(RpcError::built_in(RpcErrorCode::RecipientDisconnected, None))
+                        }
+                    },
+                    Err(_) => {
+                        self.local.pending_responses.lock().remove(&id);
+                        Err(RpcError::built_in(RpcErrorCode::ResponseTimeout, None))
+                    }
+                }
+            }
```

Comment:
Can you cleanup this code, all the results makes it hard to read

Reply:
yeah good call - i just flattened this out and its much clearer.